arch/xtensa/src/esp32/esp32_qencoder.c

```diff
@@ -0,0 +1,928 @@
+  putreg32(regval, PCNT_CTRL_REG);
+
+  priv->position = 0;
+
+  spin_unlock_irqrestore(&priv->lock, flags);
+  return OK;
+}
+
+/****************************************************************************
+ * Name: esp32_setindex
+ *
+ * Description:
+ *   Set the index pin position
+ *
+ ****************************************************************************/
+
+static int esp32_setindex(struct qe_lowerhalf_s *lower, uint32_t pos)
+{
+#ifdef CONFIG_ESP32_QENCODER_INDEX_PIN
+  struct esp32_lowerhalf_s *priv = (struct esp32_lowerhalf_s *)lower;
+  int ret = OK;
+
+  sninfo("Set QE PCNT%d the index pin positon %" PRIx32 "\n",
+         priv->config->pcntid, pos);
+  DEBUGASSERT(lower && priv->inuse);
+
+  /* Only if index pin configured */
+
+  if (priv->index_use == false)
+    {
+      snerr("ERROR: QE PCNT%d index not registered\n",
+            priv->config->pcntid);
+      ret = -EPERM;
+      goto errout;
+    }
+
+  priv->index_offset = pos;
+
+errout:
+  return ret;
+#else
+  return -ENOTTY;
+#endif
+}
+
+/****************************************************************************
+ * Name: esp32_ioctl
+ *
+ * Description:
+ *   Lower-half logic may support platform-specific ioctl commands
+ *
+ ****************************************************************************/
+
+static int esp32_ioctl(struct qe_lowerhalf_s *lower, int cmd,
+                       unsigned long arg)
+{
+  /* No ioctl commands supported */
+
+  /* TODO add an IOCTL to control the encoder pulse count prescaler */
+
+  return -ENOTTY;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: esp32_qeinitialize
+ *
+ * Description:
+ *   Initialize a quadrature encoder interface.  This function must be
+ *   called from board-specific logic.
+ *
+ * Input Parameters:
+ *   devpath - The full path to the driver to register. E.g., "/dev/qe0"
+ *   pcnt    - The PCNT number to used.  'pcnt' must be an element of
+ *             {0,1,2,3,4,5,6,7,8}
+ *
+ * Returned Value:
+ *   Zero on success; A negated errno value is returned on failure.
+ *
+ ****************************************************************************/
+
+int esp32_qeinitialize(const char *devpath, int pcnt)
+{
+  struct esp32_lowerhalf_s *priv;
+  int ret;
+
+  /* Find the pre-allocated timer state structure corresponding to this
+   * timer
+   */
+
+  priv = esp32_pcnt2lower(pcnt);
+  if (!priv)
+    {
+      snerr("ERROR: PCNT%d support not configured\n", pcnt);
+      return -ENXIO;
+    }
+
+  /* Make sure that it is available */
+
+  if (priv->inuse)
+    {
+      snerr("ERROR: PCNT%d is in-use\n", pcnt);
+      return -EBUSY;
+    }
+
+  /* Register the upper-half driver */
+
+  ret = qe_register(devpath, (struct qe_lowerhalf_s *)priv);
```

Comment:
Optional
```suggestion
  ret = qe_register(devpath, &priv->ops);
```

Reply:
Ok, I will keep it too. All other qe_register() is using that way.

Reply:
Ditto